src/utils/getBlockFromTimestamp.ts

```diff
@@ -1,165 +1,76 @@
-import type { Logger } from "pino";
-
-// estimates block number for a given timestamp using block speed
-export const estimateBlockNumber = ({
-  logger,
-  ...params
+import { BlockCache } from "../blockCache.js";
+
+export async function getBlockFromTimestamp({
+  chainId,
+  timestampInSeconds,
+  startBlock,
+  endBlock,
+  getBlockTimestamp,
+  blockCache,
 }: {
-  targetTimestamp: number;
-  startBlock: number;
-  endBlock: number;
-  startTimestamp: number;
-  endTimestamp: number;
-  logger?: Logger;
-}): number => {
-  const {
-    targetTimestamp,
-    startBlock,
-    endBlock,
-    startTimestamp,
-    endTimestamp,
-  } = params;
-
-  if (targetTimestamp < startTimestamp) {
-    logger?.trace({
-      msg: "Block estimation: Target timestamp is earlier than start timestamp",
-      diffSecs: startTimestamp - targetTimestamp,
-      params,
-    });
-  }
-  if (targetTimestamp > endTimestamp) {
-    logger?.trace({
-      msg: "Block estimation: Target timestamp is later than start timestamp",
-      params,
-    });
-  }
-  if (startTimestamp > endTimestamp) {
-    logger?.trace({
-      msg: "Block estimation: End timestamp is earlier than start timestamp",
-      params,
-    });
-  }
-  if (startBlock > endBlock) {
-    logger?.trace({
-      msg: "Block estimation: Start block is later than end block",
-      params,
-    });
-  }
-
-  const blockDistance = Math.abs(endBlock - startBlock);
-  const timeDistanceInSeconds = Math.abs(endTimestamp - startTimestamp);
-
-  // Estimate blocks per second
-  const blocksPerSecond = blockDistance / timeDistanceInSeconds;
-
-  // Now, you can use `blocksPerSecond` to adjust your block number estimation.
-  // For instance, if you know the target timestamp is X seconds away from startTimestamp, you could estimate:
-  const secondsToTarget = targetTimestamp - startTimestamp;
-  const estimatedBlocksToTarget = blocksPerSecond * secondsToTarget;
-  const estimatedBlockNumber = startBlock + Math.round(estimatedBlocksToTarget);
-
-  return estimatedBlockNumber;
-};
+  chainId: number;
+  timestampInSeconds: number;
+  startBlock: bigint;
+  endBlock: bigint;
+  getBlockTimestamp: (blockNumber: bigint) => Promise<number>;
+  blockCache?: BlockCache;
+}): Promise<bigint | null> {
```

Comment:
Does any sanity check need to be done here to warn about invalid arguments sneaking in or is done elsewhere?